lightning/src/ln/functional_tests.rs

```diff
@@ -7380,22 +7392,39 @@ fn test_data_loss_protect() {
 	}
 
 	// Check we close channel detecting A is fallen-behind
-	nodes[1].node.handle_channel_reestablish(&nodes[0].node.get_our_node_id(), &reestablish_1[0]);
-	check_closed_event!(nodes[1], 1, ClosureReason::ProcessingError { err: "Peer attempted to reestablish channel with a very old local commitment transaction".to_string() });
-	assert_eq!(check_closed_broadcast!(nodes[1], true).unwrap().data, "Peer attempted to reestablish channel with a very old local commitment transaction");
-	check_added_monitors!(nodes[1], 1);
+	// Check if we sent the warning message when we detecting that A is fallen-behind,
+	// and we give the possibility to A to  be able to recover from error.
+	expect_warning_due_wrong_channel_reestablish!(nodes[1]);
+
+	// after the waring message sent by B, we should not able to
+	// use the channel, or reconnect with success to the channel.
+	match case {
+		DataLossProtectTestCase::UseChannel => {
+			assert!(nodes[0].node.list_usable_channels().is_empty());
+		},
+		DataLossProtectTestCase::TryToReconnect => {
+			// FIXME: We sent the warning message to give the possibility to restore
+			// from the error, so we should be able to allow the reconnection here!
+			nodes[0].node.peer_connected(&nodes[1].node.get_our_node_id(), &msgs::Init { features: InitFeatures::empty(), remote_network_address: None });
+			expect_warning_due_wrong_channel_reestablish!(nodes[1]);
+		},
+	}
 
 	// Check A is able to claim to_remote output
-	let node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().clone();
-	assert_eq!(node_txn.len(), 1);
-	check_spends!(node_txn[0], chan.3);
-	assert_eq!(node_txn[0].output.len(), 2);
-	mine_transaction(&nodes[0], &node_txn[0]);
-	connect_blocks(&nodes[0], ANTI_REORG_DELAY - 1);
-	check_closed_event!(nodes[0], 1, ClosureReason::ProcessingError { err: "We have fallen behind - we have received proof that if we broadcast remote is going to claim our funds - we can\'t do any automated broadcasting".to_string() });
-	let spend_txn = check_spendable_outputs!(nodes[0], node_cfgs[0].keys_manager);
-	assert_eq!(spend_txn.len(), 1);
-	check_spends!(spend_txn[0], node_txn[0]);
+	let mut node_txn = nodes[1].tx_broadcaster.txn_broadcasted.lock().unwrap().clone();
+	// The node B should not broadcast the transaction to force close the channel!
+	assert!(node_txn.is_empty());
+	// B should now detect that there is something wrong and should force close the channel.
+	// FIXME: we sent the warning message when the peer is behind to give the possibility to restore try to fix the problem
```

Comment:
Mh! right! this was refused. I moved the FIXME a couple of lines up.

Here my claim is! If we give the possibility to the sender to force close the channel or do what they want in some cases we can fall into some malicious or very buggy code where the sender keeps reconnecting with an old data because he doesn't want to force close the channel.

Why? maybe the code tries to preserve the channel and the spec sais nothing about this case, I would like to see somethings "if the peer keep sending trash the received will force close the channel" somewhere

Reply:
A few ways to look at this:

(a) we prefer to use github issues to track things, not in-code FIXMEs or TODOs, so just as a general point would prefer to not add this kind of thing in code
(b) in general the spec shouldn't take a stance on how clients handle cases like this - its totally reasonable for a user to want the option to contact the other node operator out-of-band to avoid an on-chain closure of a big channel or when feerates are high. Its also reasonable for a node to automatically force-close based on a timer to avoid this. Its also reasonable for a node to automatically force-close right away cause the peer is buggy and screw buggy peers. Given there are many potentially correct answers here, the spec should not take a stance.
(c) on the LDK end, we really should make this something the user has more influence over, years ago our stance was to make things as automated as possible but the reality of how people want to use lightning today is somewhat in conflict with that - people get really annoyed when a channel force-closes on them. This is a topic for a separate PR, though.

Reply:
>we prefer to use github issues to track things, not in-code FIXMEs or TODOs, so just as a general point would prefer to not add this kind of thing in code

Oh right, the FIXME was added to point out this in the PR and discuss it. I think this is not really something we should catch in the ldk and also discuss in ldk because the decision of what the node should do is at the user level (from the programmer's viewpoint). So I will remove it when I will push the new fix. I'm working more in deep into the tests to make the tests clean.

>in general the spec shouldn't take a stance on how clients handle cases like this - its totally reasonable for a user to want the option to contact the other node operator out-of-band to avoid an on-chain closure of a big channel or when feerates are high. Its also reasonable for a node to automatically force-close based on a timer to avoid this. Its also reasonable for a node to automatically force-close right away cause the peer is buggy and screw buggy peers. Given there are many potentially correct answers here, the spec should not take a stance.

Right, but in the spec, I would like to see what the node should not do! e.g: Send again the wrong data! BTW this is something to discuss spec discussion, and also something useful to define also to have some global spec integration testing.